In [35]:
import root_finding as rf
import math

In [36]:
#4
def func4(x):
    return x**3 - 3*x + 1

def dfunc4(x):
    return 3*x**2 - 3

print(rf.newton(func4, dfunc4, -0.8))

-0.8
1.8740740740740753
1.6140297543160267
1.5387263232855095
1.5321383803144402
1.5320888890233573
1.5320888862379562


Interpretation of the problem:
The tangent line at -0.8 led us to the right side of root 1.532. Therefore, the function just found root 1.532, rather than the root between 0 and 1. We can't find the desired root starting with -0.8.

In [37]:
#5
def func5(x):
    return x**3 - 2*x + 2

def dfunc5(x):
    return 3*x**2 - 2

print(rf.newton(func5, dfunc5, 0))

0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0
0.0
1.0


Interpretation of the problem:
The tangent line at 0 leads us to root x_1=1. The tangent line at 1 leads us back to 0. So x oscillates between 0 and 1 and never converges. We can't find the desired root starting with 0.

In [38]:
#6 starting from -3
def func6(x):
    return math.exp(x)/(1+math.exp(x)) - 0.5

def dfunc6(x):
    return math.exp(x)/((1 + math.exp(x))**2)

print(rf.newton(func6, dfunc6, -3))

-3
7.017874927409899
-551.187496891833


OverflowError: math range error

In [39]:
#6 starting from -2
print(rf.newton(func6, dfunc6, -2))

-2
1.626860407847019
-0.8188046331667729
0.09460983291123182
-0.00014120560728900222
4.693131191828293e-13


Interpretation of the problem:
The newton's method doesn't work when x_0=-3 but works when x_0=-2. The problem with -3 is the gradient at -3 is too flat so it leads us to a point with greater x (absolute) value, where the gradient is even flatter. And then we get to a point even farther away from the root and lastly there is a math range error because the number goes out of the range. 
-2 works because the gradient is steep enough for us to converge to 0. 

In [40]:
#7
def func7(x):
    return math.sin(x) - x

def dfunc7(x):
    return math.cos(x) - 1

print(rf.newton(func7, dfunc7, -2*math.pi))

Division by zero
None


Interpretation of the problem:
The Newton method works for most x values except integer multiples of 2π, where the slope becomes 0.

In [47]:
#8a
def func8a(x):
    return abs(x)**(1/3)

def dfunc8a(x):
    #print("evaluating derivative:", x)

    if x<0:
        # 1/3 * (-x)^(-2/3)
        return -((x**2)**(-1/3))/3
    if x>0:
        return (1/3)*(x**(-2/3))
    else:
        raise Exception("Attempted to evaluate derivative at 0")

print(rf.newton(func8a, dfunc8a, 1))

1
-2.0
4.0
-7.999999999999998
15.999999999999991
-31.99999999999998
63.99999999999992
-127.99999999999986
255.99999999999957
-511.999999999999
1023.9999999999974
-2047.9999999999936
4095.9999999999845
-8191.9999999999645
16383.99999999992
-32767.999999999825
65535.99999999959
-131071.99999999907
262143.9999999979
-524287.99999999534
1048575.9999999898
-2097151.9999999772
4194303.9999999492
-8388607.999999886
16777215.999999752
-33554431.999999456
67108863.9999988
-134217727.99999744
268435455.99999452
-536870911.9999882
1073741823.9999746
-2147483647.999946
4294967295.999885
-8589934591.999756
17179869183.999481
-34359738367.998894
68719476735.99766
-137438953471.99509
274877906943.98965
-549755813887.978
1099511627775.9534
-2199023255551.902
4398046511103.794
-8796093022207.565
17592186044415.082
-35184372088830.082
70368744177659.97
-140737488355319.6
281474976710638.47
-562949953421275.4
1125899906842547.4
-2251799813685088.5
4503599627370163.5
-9007199254740298.0
1.8014398509480542

Interpretation of the problem:
The gradient is too flat so the value diverges.

In [45]:
#8b
def func8b(x):
    return abs(x)**(2/3)

def dfunc8b(x):
    #print("evaluating derivative:", x)

    if x<0:
        return 2/3*(-x)**(-1/3)
    if x>0:
        return (2/3)*(x**(-1/3))
    else:
        raise Exception("Attempted to evaluate derivative at 0")
    
print(rf.newton(func8b, dfunc8b, 1))

0.1
-0.050000000000000044
-0.12500000000000014
-0.31250000000000044
-0.7812500000000011
-1.9531250000000029
-4.882812500000008
-12.207031250000021
-30.51757812500005
-76.29394531250011
-190.73486328125026
-476.83715820312557
-1192.0928955078136
-2980.2322387695335
-7450.580596923832
-18626.451492309574
-46566.128730773926
-116415.32182693477
-291038.3045673368
-727595.7614183417
-1818989.4035458534
-4547473.508864632
-11368683.772161573
-28421709.430403918
-71054273.57600975
-177635683.94002426
-444089209.85006034
-1110223024.6251502
-2775557561.562874
-6938893903.90718
-17347234759.767937
-43368086899.419815
-108420217248.54947
-271050543121.37344
-677626357803.433
-1694065894508.581
-4235164736271.4487
-10587911840678.613
-26469779601696.508
-66174449004241.21
-165436122510602.84
-413590306276506.6
-1033975765691265.4
-2584939414228160.5
-6462348535570394.0
-1.6155871338925966e+16
-4.038967834731486e+16
-1.0097419586828704e+17
-2.5243548967071728e+17
-6.310887241767924e+17
-1.5777218

In [48]:
#8c
def func8c(x):
    return abs(x)**(4/3)

def dfunc8c(x):
    #print("evaluating derivative:", x)

    if x<0:
        return 4/3*(-x)**(1/3)
    if x>0:
        return (4/3)*(x**(1/3))
    else:
        raise Exception("Attempted to evaluate derivative at 0")
    
print(rf.newton(func8c, dfunc8c, 1))

500
125.00000000000011
31.250000000000043
7.812500000000014
1.9531250000000036
0.48828125000000067
0.12207031250000017
0.030517578125000028
0.00762939453125
0.0019073486328124983
0.00047683715820312413
0.00011920928955078082
2.9802322387695136e-05
7.450580596923767e-06
1.8626451492309384e-06
4.6566128730773354e-07
1.1641532182693312e-07
2.91038304567332e-08


Interpretation of the problem:
Only the last one converges because the numerator is smaller than the denominator. The others don't converge at all no matter how close from 0 we start. 